### PR TITLE
Staking v4 - Integration tests with custom scenarios

### PR DESCRIPTION
**Pull request by mariusmihaic**
Bug fixes
---
- Fixed bug in testing helper func `AddValidatorData` 

Refactor
---
- Moved `GetAllNodeKeys` from `systemSCs.go` to a new file to be used in tests as well
- Moved `ProcessSCOutputAccounts` from `TestNodeProcessor` to be used in staking v4 tests
- Added `baseTestMetaProcessor.go` which handles creation of a `TestMetaProcessor` regardless of `NewTestMetaProcessor`/`NewTestMetaProcessorWithCustomNodes`. Most of the functionality in this file is taken from `testMetaProcessor.go`

New tests
---
-  Added `StakingToPeer` to handle staking new nodes after staking v4 epoch, which should save new peer accounts in auction
- `TestStakingV4_UnStakeNodesWithNotEnoughFunds` which tests how nodes are unstaked at the end of the epoch in case they don't have enough funds. Before staking v4(including staking v4 init epoch), nodes should be unstaked based on prority: queue -> waiting -> eligible. After staking v4, priority should be auction -> waiting -> eligible
- `TestStakingV4_StakeNewNodes` tests functionality of staking new nodes after staking v4(they should be sent to auction)

### CODE REVIEW

File: integrationTests/vm/staking/baseTestMetaProcessor.go [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+	bootstrapComponents factory.BootstrapComponentsHolder,
+	statusComponents factory.StatusComponentsHolder,
+	stateComponents factory.StateComponentsHandler,
+	nc nodesCoordinator.NodesCoordinator,
+	maxNodesConfig []config.MaxNodesChangeConfig,
+	queue [][]byte,
+) *TestMetaProcessor {
+	gasScheduleNotifier := createGasScheduleNotifier()
+	blockChainHook := createBlockChainHook(
+		dataComponents, coreComponents,
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/baseTestMetaProcessor.go:85]: Right, missed it;
Added new line

File: epochStart/metachain/systemSCs.go [hunk 2]
```diff
@@ -320,7 +320,7 @@ func calcNormRand(randomness []byte, expectedLen int) []byte {
 	randLen := len(rand)
 
 	if expectedLen > randLen {
-		repeatedCt := expectedLen/randLen + 1
+		repeatedCt := expectedLen/randLen + 1 // todo: fix possible div by 0
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:323]: `expectedLen/randLen` is the first operation being executed; `randLen` could be 0 => possible div by 0

File: integrationTests/vm/staking/stakingV4_test.go [hunk 3]
```diff
@@ -65,7 +73,37 @@ func requireMapDoesNotContain(t *testing.T, m map[uint32][][]byte, s [][]byte) {
 	}
 }
 
-// TODO: Staking v4: more tests to check exactly which nodes have been selected/unselected from previous nodes config auction
+func remove(s [][]byte, elem []byte) [][]byte {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/stakingV4_test.go:76]: I think this name is a bit too lengthy and maybe not worth describing its behavior for a test. Added a comm to highlight its behavior. Also, renamed `s` -> `slice`

File: epochStart/metachain/systemSCs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:323]: Left a `//todo` there so I can fix it in the next PR in which I will create a separate component for auction selection.
This is going to be fixed in the next PR, along with the new subcomp, so we don't "clutter" this PR with anymore changes.

**Review by sasurobert — APPROVED** [review]

**Pull request merged**